File: clients/android/app/src/main/java/com/v0id/wallet/core/Encryption.kt
package com.v0id.wallet.core

import org.bouncycastle.crypto.modes.ChaCha20Poly1305
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.math.ec.rfc7748.X25519
import java.math.BigInteger
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * 端到端加密私信：ed25519 → x25519 → ECDH 共享密钥 → XChaCha20-Poly1305 认证加密。
 * 必须与 packages/core/src/crypto.ts（@noble/curves + @noble/ciphers）逐字节互通，否则全网解不开。
 * 自检向量见 CLIENT-PROTOCOL §8.6 / §9（GoldVectorTest 用同一组向量对齐）。
 *
 * 依赖拆解（零新依赖，bcprov 已是依赖）：
 *   • x25519 标量乘 → BouncyCastle X25519（rfc7748）。
 *   • IETF ChaCha20-Poly1305 → BouncyCastle ChaCha20Poly1305。
 *   • ed25519 公钥 → Montgomery u 坐标的域运算 → java.math.BigInteger（mod p=2^255-19）。
 *   • HChaCha20 子密钥派生（XChaCha 的前半段）→ 本文件手写。
 */
object Encryption {
    // p = 2^255 - 19（GF 域素数）
    private val P: BigInteger = BigInteger.TWO.pow(255).subtract(BigInteger.valueOf(19))

    // ───── 对外 API ─────

    fun isEncryptedMemo(memo: String): Boolean = memo.startsWith(ENC_PREFIX)

    /** 加密一段明文给收件人（用发送方自己的种子私钥）。返回 `ENC|<hex>`，直接当 memo 上链。失败返回 null。 */
    fun encryptMemo(plaintext: String, recipientAddress: String, senderSeed: ByteArray): String? {
        val key = sharedKey(senderSeed, recipientAddress) ?: return null
        val nonce = ByteArray(24).also { SecureRandom().nextBytes(it) }
        val sealed = try {
            aeadSeal(plaintext.toByteArray(Charsets.UTF_8), key, nonce)
        } catch (_: Exception) {
            return null
        }
        return ENC_PREFIX + (nonce + sealed).toHex()
    }

    /**
     * 解密一条 `ENC|` 私信。otherPartyAddress = 对方地址（我是收件人→填发件人；我是发件人→填收件人）。
     * 失败（非本人 / 被篡改 / 格式坏）返回 null。ECDH 对称 → 发件人也能解自己发的。
     */
    fun decryptMemo(memo: String, otherPartyAddress: String, mySeed: ByteArray): String? {
        if (!memo.startsWith(ENC_PREFIX)) return null
        val blob = try {
            memo.substring(ENC_PREFIX.length).hexToBytes()
        } catch (_: Exception) {
            return null
        }
        if (blob.size < 24 + 16) return null
        val nonce = blob.copyOfRange(0, 24)
        val ctTag = blob.copyOfRange(24, blob.size)
        val key = sharedKey(mySeed, otherPartyAddress) ?: return null
        return try {
            String(aeadOpen(ctTag, key, nonce), Charsets.UTF_8)
        } catch (_: Exception) {
            null
        }
    }

    // ───── 共享密钥：我的 ed25519 种子 × 对方地址(ed25519 公钥) → 32 字节对称密钥 ─────
    fun sharedKey(mySeed: ByteArray, otherAddress: String): ByteArray? {
        val pubHex = if (otherAddress.startsWith("0x")) otherAddress.substring(2) else otherAddress
        val otherPub = try {
            pubHex.hexToBytes()
        } catch (_: Exception) {
            return null
        }
        if (otherPub.size != 32) return null
        val xPriv = x25519PrivFromSeed(mySeed)            // clamp(sha512(seed)[:32])
        val otherU = montgomeryUFromEdPub(otherPub)       // ed25519 公钥 → Montgomery u
        val out = ByteArray(32)
        // X25519 内部会再 clamp（幂等），传入已 clamp 的私钥结果不变；输出 = 原始 ECDH 32 字节，无 KDF。
        X25519.calculateAgreement(xPriv, 0, otherU, 0, out, 0)
        return out
    }

    /** ed25519 私钥种子 → x25519 私钥标量 = clamp(SHA-512(seed)[0:32])。 */
    fun x25519PrivFromSeed(seed: ByteArray): ByteArray {
        val h = MessageDigest.getInstance("SHA-512").digest(seed).copyOfRange(0, 32)
        h[0] = (h[0].toInt() and 248).toByte()
        h[31] = (h[31].toInt() and 127).toByte()
        h[31] = (h[31].toInt() or 64).toByte()
        return h
    }

    /** ed25519 公钥(32B) → Montgomery u = (1+y)/(1-y) mod p，编码为 32 字节小端。 */
    fun montgomeryUFromEdPub(pub: ByteArray): ByteArray {
        // y = 小端(pub)，清掉最高位（符号位 bit255）后 mod p
        val yBytes = pub.copyOf()
        yBytes[31] = (yBytes[31].toInt() and 0x7f).toByte()
        val y = leToBigInteger(yBytes).mod(P)
        val onePlusY = BigInteger.ONE.add(y).mod(P)
        val oneMinusY = BigInteger.ONE.subtract(y).mod(P)
        val inv = oneMinusY.modInverse(P)
        val u = onePlusY.multiply(inv).mod(P)
        return bigIntegerToLe(u, 32)
    }

    // ───── XChaCha20-Poly1305 = HChaCha20 派生子密钥 + IETF ChaCha20-Poly1305 ─────
    fun aeadSeal(plaintext: ByteArray, key: ByteArray, xnonce: ByteArray): ByteArray {
        val (subkey, ietf) = xchachaSetup(key, xnonce)
        val cipher = ChaCha20Poly1305()
        cipher.init(true, AEADParameters(KeyParameter(subkey), 128, ietf))
        val out = ByteArray(cipher.getOutputSize(plaintext.size))
        val len = cipher.processBytes(plaintext, 0, plaintext.size, out, 0)
        cipher.doFinal(out, len)
        return out  // 密文 ‖ 16 字节 tag（noble 格式）
    }

    fun aeadOpen(ctTag: ByteArray, key: ByteArray, xnonce: ByteArray): ByteArray {
        val (subkey, ietf) = xchachaSetup(key, xnonce)
        val cipher = ChaCha20Poly1305()
        cipher.init(false, AEADParameters(KeyParameter(subkey), 128, ietf))
        val out = ByteArray(cipher.getOutputSize(ctTag.size))
        val len = cipher.processBytes(ctTag, 0, ctTag.size, out, 0)
        cipher.doFinal(out, len)  // tag 不符会抛异常
        return out
    }

    /** 子密钥 = HChaCha20(key, xnonce[0:16])；IETF nonce = 0x00000000 ‖ xnonce[16:24]（12 字节）。 */
    private fun xchachaSetup(key: ByteArray, xnonce: ByteArray): Pair<ByteArray, ByteArray> {
        val subkey = hchacha20(key, xnonce.copyOfRange(0, 16))
        val ietf = ByteArray(12)
        for (i in 0 until 8) ietf[4 + i] = xnonce[16 + i]
        return subkey to ietf
    }

    // ───── HChaCha20：从 256-bit key + 128-bit nonce 派生 256-bit 子密钥 ─────
    fun hchacha20(key: ByteArray, nonce16: ByteArray): ByteArray {
        fun le32(b: ByteArray, o: Int): Int =
            (b[o].toInt() and 0xff) or
                ((b[o + 1].toInt() and 0xff) shl 8) or
                ((b[o + 2].toInt() and 0xff) shl 16) or
                ((b[o + 3].toInt() and 0xff) shl 24)

        val s = IntArray(16)
        s[0] = 0x61707865; s[1] = 0x3320646e; s[2] = 0x79622d32; s[3] = 0x6b206574
        for (i in 0 until 8) s[4 + i] = le32(key, i * 4)
        for (i in 0 until 4) s[12 + i] = le32(nonce16, i * 4)

        fun rotl(x: Int, n: Int): Int = (x shl n) or (x ushr (32 - n))
        fun qr(a: Int, b: Int, c: Int, d: Int) {
            s[a] += s[b]; s[d] = rotl(s[d] xor s[a], 16)
            s[c] += s[d]; s[b] = rotl(s[b] xor s[c], 12)
            s[a] += s[b]; s[d] = rotl(s[d] xor s[a], 8)
            s[c] += s[d]; s[b] = rotl(s[b] xor s[c], 7)
        }
        repeat(10) {
            qr(0, 4, 8, 12); qr(1, 5, 9, 13); qr(2, 6, 10, 14); qr(3, 7, 11, 15)
            qr(0, 5, 10, 15); qr(1, 6, 11, 12); qr(2, 7, 8, 13); qr(3, 4, 9, 14)
        }
        val out = ByteArray(32)
        fun put(v: Int, o: Int) {
            out[o] = (v and 0xff).toByte()
            out[o + 1] = ((v ushr 8) and 0xff).toByte()
            out[o + 2] = ((v ushr 16) and 0xff).toByte()
            out[o + 3] = ((v ushr 24) and 0xff).toByte()
        }
        put(s[0], 0); put(s[1], 4); put(s[2], 8); put(s[3], 12)
        put(s[12], 16); put(s[13], 20); put(s[14], 24); put(s[15], 28)
        return out
    }

    // ───── 小端 ↔ BigInteger（BigInteger 是大端，需翻转字节序）─────
    private fun leToBigInteger(le: ByteArray): BigInteger {
        val be = ByteArray(le.size)
        for (i in le.indices) be[i] = le[le.size - 1 - i]
        return BigInteger(1, be)  // 1 = 正号，避免最高位被当符号
    }

    private fun bigIntegerToLe(v: BigInteger, len: Int): ByteArray {
        val out = ByteArray(len)
        var x = v
        val mask = BigInteger.valueOf(0xff)
        for (i in 0 until len) {
            out[i] = x.and(mask).toInt().toByte()
            x = x.shiftRight(8)
        }
        return out
    }
}

File: clients/macos/V0idKit/Sources/V0idKit/Encryption.swift
// 端到端加密私信：ed25519 → x25519 → ECDH 共享密钥 → XChaCha20-Poly1305 认证加密。
// 必须与 packages/core/src/crypto.ts（@noble/curves + @noble/ciphers）逐字节互通，否则全网解不开。
// 自检向量见 CLIENT-PROTOCOL §8.6 / §9（EncryptionTests 用同一组向量对齐）。
//
// 依赖拆解（零新依赖）：
//   • x25519 标量乘 + IETF ChaCha20-Poly1305 → Apple CryptoKit（系统框架）。
//   • ed25519 公钥 → Montgomery u 坐标的域运算 → 本文件内 radix-2^16 GF(2^255-19)
//     （TweetNaCl 风格的经典实现，公有领域；见 THIRD-PARTY-NOTICES）。
//   • HChaCha20 子密钥派生（XChaCha 的前半段）→ 本文件手写。
import Foundation
import CryptoKit

public enum Encryption {
    // ---- 对外 API ----

    public static func isEncryptedMemo(_ memo: String) -> Bool { memo.hasPrefix(Config.encPrefix) }

    /// 加密一段明文给收件人（用发送方自己的种子私钥）。返回 `ENC|<hex>`，直接当 memo 上链。失败返回 nil。
    public static func encryptMemo(_ plaintext: String, recipientAddress: String, senderSeed: Data) -> String? {
        guard let key = sharedKey(mySeed: senderSeed, otherAddress: recipientAddress) else { return nil }
        var nonce = [UInt8](repeating: 0, count: 24)
        for i in 0..<24 { nonce[i] = UInt8.random(in: 0...255) }
        guard let sealed = try? aeadSeal(plaintext: Array(plaintext.utf8), key: key, xnonce: nonce) else { return nil }
        return Config.encPrefix + Hex.encode(nonce + sealed)
    }

    /// 解密一条 `ENC|` 私信。otherPartyAddress = 对方地址（我是收件人→填发件人；我是发件人→填收件人）。
    /// 失败（非本人 / 被篡改 / 格式坏）返回 nil。ECDH 对称 → 发件人也能解自己发的。
    public static func decryptMemo(_ memo: String, otherPartyAddress: String, mySeed: Data) -> String? {
        guard memo.hasPrefix(Config.encPrefix) else { return nil }
        guard let blob = Hex.decode(String(memo.dropFirst(Config.encPrefix.count))), blob.count >= 24 + 16 else { return nil }
        let nonce = Array(blob[0..<24])
        let ctTag = Array(blob[24...])
        guard let key = sharedKey(mySeed: mySeed, otherAddress: otherPartyAddress),
              let pt = try? aeadOpen(ctTag: ctTag, key: key, xnonce: nonce) else { return nil }
        return String(decoding: pt, as: UTF8.self)
    }

    // ---- 共享密钥：我的 ed25519 种子 × 对方地址(ed25519 公钥) → 32 字节对称密钥 ----
    static func sharedKey(mySeed: Data, otherAddress: String) -> [UInt8]? {
        let pubHex = otherAddress.hasPrefix("0x") ? String(otherAddress.dropFirst(2)) : otherAddress
        guard let otherPub = Hex.decode(pubHex), otherPub.count == 32 else { return nil }
        let xPriv = x25519PrivFromSeed(Array(mySeed))            // clamp(sha512(seed)[:32])
        let otherU = montgomeryUFromEdPub(otherPub)             // ed25519 公钥 → Montgomery u
        guard
            let priv = try? Curve25519.KeyAgreement.PrivateKey(rawRepresentation: Data(xPriv)),
            let pub = try? Curve25519.KeyAgreement.PublicKey(rawRepresentation: Data(otherU)),
            let secret = try? priv.sharedSecretFromKeyAgreement(with: pub)
        else { return nil }
        return secret.withUnsafeBytes { Array($0) }              // 原始 x25519 ECDH 结果（32 字节，无 KDF）
    }

    /// ed25519 私钥种子 → x25519 私钥标量 = clamp(SHA-512(seed)[0:32])。
    static func x25519PrivFromSeed(_ seed: [UInt8]) -> [UInt8] {
        var h = Array(SHA512.hash(data: Data(seed)).prefix(32))
        h[0] &= 248
        h[31] &= 127
        h[31] |= 64
        return h
    }

    /// ed25519 公钥(32B) → Montgomery u = (1+y)/(1-y) mod p，编码为 32 字节小端。
    static func montgomeryUFromEdPub(_ pub: [UInt8]) -> [UInt8] {
        let y = GF.unpack(pub)                 // 取低 255 位作为 y（清掉 x 符号位）
        let onePlusY = GF.add(GF.one, y)
        let oneMinusY = GF.sub(GF.one, y)
        let inv = GF.inv(oneMinusY)
        let u = GF.mul(onePlusY, inv)
        return GF.pack(u)
    }

    // ---- XChaCha20-Poly1305 = HChaCha20 派生子密钥 + IETF ChaCha20-Poly1305 ----
    static func aeadSeal(plaintext: [UInt8], key: [UInt8], xnonce: [UInt8]) throws -> [UInt8] {
        let (subkey, ietf) = xchachaSetup(key: key, xnonce: xnonce)
        let box = try ChaChaPoly.seal(Data(plaintext),
                                      using: SymmetricKey(data: Data(subkey)),
                                      nonce: try ChaChaPoly.Nonce(data: Data(ietf)))
        return Array(box.ciphertext) + Array(box.tag)            // 密文 ‖ 16 字节 tag（noble 格式）
    }

    static func aeadOpen(ctTag: [UInt8], key: [UInt8], xnonce: [UInt8]) throws -> [UInt8] {
        let (subkey, ietf) = xchachaSetup(key: key, xnonce: xnonce)
        let ct = Array(ctTag[0..<(ctTag.count - 16)])
        let tag = Array(ctTag[(ctTag.count - 16)...])
        let box = try ChaChaPoly.SealedBox(nonce: try ChaChaPoly.Nonce(data: Data(ietf)),
                                           ciphertext: Data(ct), tag: Data(tag))
        return Array(try ChaChaPoly.open(box, using: SymmetricKey(data: Data(subkey))))
    }

    /// 子密钥 = HChaCha20(key, xnonce[0:16])；IETF nonce = 0x00000000 ‖ xnonce[16:24]（12 字节）。
    private static func xchachaSetup(key: [UInt8], xnonce: [UInt8]) -> (subkey: [UInt8], ietf: [UInt8]) {
        let subkey = hchacha20(key: key, nonce16: Array(xnonce[0..<16]))
        var ietf = [UInt8](repeating: 0, count: 12)
        for i in 0..<8 { ietf[4 + i] = xnonce[16 + i] }
        return (subkey, ietf)
    }

    // ---- HChaCha20：从 256-bit key + 128-bit nonce 派生 256-bit 子密钥 ----
    static func hchacha20(key: [UInt8], nonce16: [UInt8]) -> [UInt8] {
        func le32(_ b: [UInt8], _ o: Int) -> UInt32 {
            UInt32(b[o]) | (UInt32(b[o + 1]) << 8) | (UInt32(b[o + 2]) << 16) | (UInt32(b[o + 3]) << 24)
        }
        var s = [UInt32](repeating: 0, count: 16)
        s[0] = 0x6170_7865; s[1] = 0x3320_646e; s[2] = 0x7962_2d32; s[3] = 0x6b20_6574
        for i in 0..<8 { s[4 + i] = le32(key, i * 4) }
        for i in 0..<4 { s[12 + i] = le32(nonce16, i * 4) }
        func rotl(_ x: UInt32, _ n: UInt32) -> UInt32 { (x << n) | (x >> (32 - n)) }
        func qr(_ a: Int, _ b: Int, _ c: Int, _ d: Int) {
            s[a] = s[a] &+ s[b]; s[d] = rotl(s[d] ^ s[a], 16)
            s[c] = s[c] &+ s[d]; s[b] = rotl(s[b] ^ s[c], 12)
            s[a] = s[a] &+ s[b]; s[d] = rotl(s[d] ^ s[a], 8)
            s[c] = s[c] &+ s[d]; s[b] = rotl(s[b] ^ s[c], 7)
        }
        for _ in 0..<10 {
            qr(0, 4, 8, 12); qr(1, 5, 9, 13); qr(2, 6, 10, 14); qr(3, 7, 11, 15)
            qr(0, 5, 10, 15); qr(1, 6, 11, 12); qr(2, 7, 8, 13); qr(3, 4, 9, 14)
        }
        var out = [UInt8](repeating: 0, count: 32)
        func put(_ v: UInt32, _ o: Int) {
            out[o] = UInt8(v & 0xff); out[o + 1] = UInt8((v >> 8) & 0xff)
            out[o + 2] = UInt8((v >> 16) & 0xff); out[o + 3] = UInt8((v >> 24) & 0xff)
        }
        put(s[0], 0); put(s[1], 4); put(s[2], 8); put(s[3], 12)
        put(s[12], 16); put(s[13], 20); put(s[14], 24); put(s[15], 28)
        return out
    }
}

// ---- GF(2^255-19) 有限域：radix-2^16（16 个 Int64 limb）经典实现（TweetNaCl 风格，公有领域）----
// 仅用于把 ed25519 公钥转 Montgomery u 坐标（一次模逆 + 两次乘）。已用金标准向量逐字节验证。
enum GF {
    typealias Elem = [Int64]   // 16 个 limb，radix 2^16
    static let one: Elem = { var g = Elem(repeating: 0, count: 16); g[0] = 1; return g }()

    static func add(_ a: Elem, _ b: Elem) -> Elem { (0..<16).map { a[$0] + b[$0] } }
    static func sub(_ a: Elem, _ b: Elem) -> Elem { (0..<16).map { a[$0] - b[$0] } }

    /// 进位归一（把每个 limb 收回 16 位，2^256 ≡ 38 折叠到低位）。
    static func carry(_ o: inout Elem) {
        var c: Int64 = 0
        for i in 0..<16 {
            o[i] += 1 << 16
            c = o[i] >> 16
            if i < 15 { o[i + 1] += c - 1 } else { o[0] += 38 * (c - 1) }
            o[i] -= c << 16
        }
    }

    static func mul(_ a: Elem, _ b: Elem) -> Elem {
        var t = [Int64](repeating: 0, count: 31)
        for i in 0..<16 { for j in 0..<16 { t[i + j] += a[i] * b[j] } }
        for i in 0..<15 { t[i] += 38 * t[i + 16] }
        var o = Array(t[0..<16])
        carry(&o); carry(&o)
        return o
    }

    static func sq(_ a: Elem) -> Elem { mul(a, a) }

    /// 模逆：a^(p-2)，p-2 的位模式即 TweetNaCl inv25519 的循环（除 bit 2 / 4 外都乘 a）。
    static func inv(_ i: Elem) -> Elem {
        var c = i
        for a in stride(from: 253, through: 0, by: -1) {
            c = sq(c)
            if a != 2 && a != 4 { c = mul(c, i) }
        }
        return c
    }

    /// 32 字节小端 → 域元素（清掉最高位，作为 y 坐标）。
    static func unpack(_ n: [UInt8]) -> Elem {
        var o = Elem(repeating: 0, count: 16)
        for i in 0..<16 { o[i] = Int64(n[2 * i]) + (Int64(n[2 * i + 1]) << 8) }
        o[15] &= 0x7fff
        return o
    }

    /// 域元素 → 32 字节小端（完全约简到 [0, p)）。
    static func pack(_ n: Elem) -> [UInt8] {
        var t = n
        carry(&t); carry(&t); carry(&t)
        for _ in 0..<2 {
            var m = Elem(repeating: 0, count: 16)
            m[0] = t[0] - 0xffed
            for i in 1..<15 {
                m[i] = t[i] - 0xffff - ((m[i - 1] >> 16) & 1)
                m[i - 1] &= 0xffff
            }
            m[15] = t[15] - 0x7fff - ((m[14] >> 16) & 1)
            let b = (m[15] >> 16) & 1
            m[14] &= 0xffff
            if (1 - b) == 1 { t = m }   // 未下溢（t ≥ p）→ 用减去 p 后的 m
        }
        var o = [UInt8](repeating: 0, count: 32)
        for i in 0..<16 {
            o[2 * i] = UInt8(t[i] & 0xff)
            o[2 * i + 1] = UInt8((t[i] >> 8) & 0xff)
        }
        return o
    }
}
